File: 0047-permutations-ii/0047-permutations-ii.java
class Solution {

    public void swap(int i, int j, int[] nums){
        int t = nums[i];
        nums[i] = nums[j];
        nums[j] = t;
    }

    public void solve(int[] nums, int idx, Set<List<Integer>> result){
        if(idx >= nums.length){
            result.add(Arrays.stream(nums)
                         .boxed()
                         .collect(Collectors.toList()));
            return;
        }

        for(int i=idx; i<nums.length; i++){
            swap(idx, i, nums);
            solve(nums, idx+1, result);
            swap(idx, i, nums);
        }
    }

    public List<List<Integer>> permuteUnique(int[] nums) {
        Set<List<Integer>> result = new HashSet<>();
        solve(nums, 0, result);
        return result.stream().collect(Collectors.toList());
    }
}

File: 0047-permutations-ii/0047-permutations-ii.cpp
class Solution {
public:
    void solve(int idx, vector<int> &nums, vector<vector<int>> &result){
        if(idx >= nums.size()){
            result.push_back(nums);
        }

        unordered_set<int> uniqueSet;
        for(int i = idx; i < nums.size(); i++) {
            if(uniqueSet.count(nums[i]) > 0) {
                continue;
            }
            uniqueSet.insert(nums[i]);
            swap(nums[i], nums[idx]);
            solve(idx+1, nums, result);
            swap(nums[i], nums[idx]);
        }
    }

    vector<vector<int>> permuteUnique(vector<int>& nums) {
        vector<vector<int>> result;
        solve(0, nums, result);
        return result;
    }
};